src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,416 @@
+
+            var targetNode = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true)
+                .GetAncestorsOrThis<ExpressionSyntax>().FirstOrDefault();
+            if (targetNode == null)
+                return;
+
+            var hasSolution = TryGetTargetTypeInfo(
+                semanticModel, diagnostic.Id, targetNode, cancellationToken,
+                out var nodeType, out var potentialConversionTypes);
+            if (!hasSolution)
+            {
+                return;
+            }
+
+            if (potentialConversionTypes.Length == 1)
+            {
+                context.RegisterCodeFix(new MyCodeAction(
+                    CSharpFeaturesResources.Add_explicit_cast,
+                    c => FixAsync(context.Document, context.Diagnostics.First(), c)),
+                    context.Diagnostics);
+            }
+            else
+            {
+                var actions = ArrayBuilder<CodeAction>.GetInstance();
+
+                // MaximumConversionOptions: we show at most [MaximumConversionOptions] options for this code fixer
+                for (var i = 0; i < Math.Min(MaximumConversionOptions, potentialConversionTypes.Length); i++)
+                {
+                    var convType = potentialConversionTypes[i];
+                    actions.Add(new MyCodeAction(
+                        string.Format(
+                            CSharpFeaturesResources.Convert_type_to_0,
+                            convType.ToMinimalDisplayString(semanticModel, context.Span.Start)),
+                        _ => ApplySingleConversionToDocumentAsync(document, ApplyFix(root, targetNode, convType))));
+                }
+
+                if (potentialConversionTypes.Length > MaximumConversionOptions)
+                {
+                    // If the number of potential conversion types is larger than options we could show, report telemetry
+                    Logger.Log(FunctionId.CodeFixes_AddExplicitCast,
+                        KeyValueLogMessage.Create(m =>
+                        {
+                            m["NumberOfCandidates"] = potentialConversionTypes.Length;
+                        }));
+                }
+
+                context.RegisterCodeFix(new CodeAction.CodeActionWithNestedActions(
+                    CSharpFeaturesResources.Add_explicit_cast,
+                    actions.ToImmutableAndFree(), isInlinable: false),
+                    context.Diagnostics);
+            }
+        }
+
+        private static SyntaxNode ApplyFix(SyntaxNode currentRoot, ExpressionSyntax targetNode, ITypeSymbol conversionType)
+        {
+            // TODO: castExpression.WithAdditionalAnnotations(Simplifier.Annotation) 
+            // - the Simplifier doesn't remove the redundant cast from the expression
+            // Issue link: https://github.com/dotnet/roslyn/issues/41500
```

Comment:
i don't understand this comment.  the TODO lists what hte code actually is.

Reply:
I remove the actual code, here just means `castExpression.WithAdditionalAnnotations(Simplifier.Annotation)` doesn't work properly :)